File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

function permutations(input ,output,array){
    
      if (input.length === 0) {
            array.push(output); 
            return;
        }
        for (let i = 0; i < input.length; i++) {
            
            let newInput = [...input.slice(0, i), ...input.slice(i + 1)];
            let newOutput = [...output, input[i]];
            
            permutations(newInput, newOutput,array);
        }
  return array;

}

var permute = function(nums) {
let array =[];
let output="";
 let result = permutations(nums,output,array);
 return result;
    
};

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> result;
    void solve(vector<int>& nums,vector<int>& temp, vector<bool>& used) {

        if(temp.size() == nums.size()) {
            result.push_back(temp);
            return;
        }

        for(int i = 0; i < nums.size(); i++) {

            if(!used[i]) {
                used[i] = true;
                temp.push_back(nums[i]);

                solve(nums,  temp, used);

                temp.pop_back();     
                used[i] = false;     
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {

        vector<int> temp;
        vector<bool> used(nums.size(), false);

        solve(nums,  temp, used);
        return result;
    }
};